Sources/GRPC/GRPCClientConnection.swift

```diff
@@ -171,7 +153,163 @@ open class GRPCClientConnection {
   }
 }
 
+// MARK:- Configuration structures
+
+/// A target to connect to.
+public enum ConnectionTarget {
+  /// The host and port.
+  case hostAndPort(String, Int)
+  /// The path of a Unix domain socket.
+  case unixDomainSocket(String)
+  /// A NIO socket address.
+  case socketAddress(SocketAddress)
+
+  var host: String? {
+    guard case .hostAndPort(let host, _) = self else {
+      return nil
+    }
+    return host
+  }
+}
+
 extension GRPCClientConnection {
+  /// The configuration for a connection.
+  public struct Configuration {
+    /// The target to connect to.
+    public var target: ConnectionTarget
+
+    /// The event loop group to run the connection on.
+    public var eventLoopGroup: EventLoopGroup
+
+    /// An error delegate which is called when errors are caught. Provided delegates **must not
+    /// maintain a strong reference to this `GRPCClientConnection`**. Doing so will cause a retain
+    /// cycle.
+    public var errorDelegate: ClientErrorDelegate?
+
+    /// TLS configuration for this connection. `nil` if TLS is not desired.
+    public var tlsConfiguration: TLSConfiguration?
+
+    /// The HTTP protocol used for this connection.
+    public var httpProtocol: HTTP2ToHTTP1ClientCodec.HTTPProtocol {
+      return self.tlsConfiguration == nil ? .http : .https
+    }
+
+    /// Create a `Configuration` with some pre-defined defaults.
+    ///
+    /// - Parameter target: The target to connect to.
+    /// - Parameter eventLoopGroup: The event loop group to run the connection on.
+    /// - Parameter errorDelegate: The error delegate, defaulting to a delegate which will log only
+    ///     on debug builds.
+    /// - Parameter tlsConfiguration: TLS configuration, defaulting to `nil`.
+    public init(
+      target: ConnectionTarget,
+      eventLoopGroup: EventLoopGroup,
+      errorDelegate: ClientErrorDelegate? = DebugOnlyLoggingClientErrorDelegate.shared,
+      tlsConfiguration: TLSConfiguration? = nil
+      ) {
+      self.target = target
+      self.eventLoopGroup = eventLoopGroup
+      self.errorDelegate = errorDelegate
+      self.tlsConfiguration = tlsConfiguration
+    }
+  }
+
+  /// The TLS configuration for a connection.
+  public struct TLSConfiguration {
```

Comment:
Should we provide a "default" initializer similar to the one you created in the Echo example?

Reply:
I'm not sure what you mean here?

Reply:
Essentially moving the `makeClientSSLContext` from the example into the framework :-)

Reply:
Oh I see, I'm going to hold off on this for now if you don't mind because it just occurred to me that setting up SSL for NIOTS is different, so no doubt this configuration will have to change anyway.